index/segment/mem/trigram_terms_dict.go

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"regexp/syntax"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+
+	"github.com/google/codesearch/index"
+)
+
+var (
+	// sentinelTrigram is stored for every field so that we can perform match all lookups.
+	sentinelTrigram = ""
+)
+
+// The trigram terms dictionary works by breaking down the value of a field into its constiuent
+// trigrams and storing each trigram in a simple dictionary. For example, given a field
+// (name: "foo", value: "fizzbuzz") and DocID `i` we
+//   (1) Compute all trigrams for the given value. In this case:
+//         fiz, izz, zzb, zbu, buz, uzz
+//   (2) For each trigram `t` created in step 1, store the entry (value: `t`, docID: `i`) in the
+//       postings list for the field name "foo".
+//
+
+// trigramsTermsDict uses trigrams in a two-level map to model a terms dictionary.
+// i.e. fieldName > trigram(fieldValue) -> postingsList
+type trigramTermsDictionary struct {
+	opts Options
+
+	fields struct {
```

Comment:
I don't follow the thought process here:
- would be more efficient to return the candidate docIDs from `Fetch` - that way, we can merge across multiple fields before performing validation.
- document -> fields is many one, should the map value be `[]doc.Field`?

Reply:
For the second point, yea, it should be `[]doc.Field`, I was mistaken because the `Insert` method only takes a single field, will update.

For the first point, if we just return the candidate docIDs from `Fetch` then we can return false positives. I wanted to avoid this since then the index would always have to check for false positives in the docIDs returned from a segment, even though some segments don't return any false positives.